lib/streamlit/errors.py

```diff
@@ -133,3 +137,227 @@ def __init__(self, module_name, *args):
             "installed."
         )
         super().__init__(message, *args)
+
+
+class LocalizableStreamlitException(StreamlitAPIException):
+    def __init__(self, message: str, **kwargs):
+        exception_name = type(self).__name__
+        exception_message = gettext.gettext(exception_name)
+        if exception_message == exception_name:
+            # gettext returns the message ID if there is no translation.
+            # In this case, we want to use the message provided in the constructor.
+            exception_message = message
+        super().__init__((exception_message).format(**kwargs))
+        self._exec_kwargs = kwargs
+
+    @property
+    def exec_kwargs(self) -> dict[str, Any]:
+        return self._exec_kwargs
+
+
+# st.number_input
+class StreamlitNumberInputDifferentTypesError(LocalizableStreamlitException):
+    """Exception raised mixing floats and ints in st.number_input."""
+
+    def __init__(
+        self,
+        value: int | float,
+        min_value: int | float,
+        max_value: int | float,
+        step: int | float,
+    ):
+        error_message = "All numerical arguments must be of the same type."
+
+        if value:
+            value_type = type(value).__name__
+            error_message += f"\n`value` has {value_type} type."
+
+        if min_value:
+            min_value_type = type(min_value).__name__
+            error_message += f"\n`min_value` has {min_value_type} type."
+
+        if max_value:
+            max_value_type = type(max_value).__name__
+            error_message += f"\n`max_value` has {max_value_type} type."
+
+        if step:
+            step_type = type(step).__name__
+            error_message += f"\n`step` has {step_type} type."
+
+        super().__init__(error_message)
+
+
+class StreamlitInputInvalidValueError(LocalizableStreamlitException):
+    """Exception raised when the `min_value` is greater or the `max_value` is less than the `value` in `st.number_input`."""
+
+    def __init__(
+        self, value: int | float, min_value: int | float, max_value: int | float
+    ):
+        error_message = None
+
+        if min_value is not None and value < min_value:
+            error_message = (
+                f"The `value` {value} is less than the `min_value` {min_value}."
+            )
+        elif max_value is not None and value > max_value:
+            error_message = (
+                f"The `value` {value} is greater than the `max_value` {max_value}."
```

Comment:
Related to the comment: https://github.com/streamlit/streamlit/pull/9318/files#r1727107502 , it's probably important that we keep a 1-to-1 mapping from exception message to exception type. So, either we can write the message more generic to cover multiple cases with the same template, or we can split up into two exceptions.

Reply:
You got it!